Framework/Muon/src/ApplyMuonDetectorGrouping.cpp

```diff
@@ -0,0 +1,503 @@
+      inputGroup->addWorkspace(ws);
+    }
+    alg.setProperty("SummedPeriodSet", options.summedPeriods);
+    alg.setProperty("SubtractedPeriodSet", options.subtractedPeriods);
+  } else if (auto ws = boost::dynamic_pointer_cast<MatrixWorkspace>(inputWS)) {
+    // Put this single WS into a group and set it as the input property
+    inputGroup->addWorkspace(ws);
+    alg.setProperty("SummedPeriodSet", "1");
+  } else {
+    throw std::runtime_error("Cannot create workspace: workspace must be "
+                             "MatrixWorkspace or WorkspaceGroup.");
+  }
+  alg.setProperty("InputWorkspace", inputGroup);
+}
+
+/**
+ * Set time properties according to the given options. For use with
+ * MuonProcess.
+ */
+void ApplyMuonDetectorGrouping::setMuonProcessAlgorithmTimeProperties(
+    IAlgorithm &alg, const Muon::AnalysisOptions &options) const {
+  alg.setProperty("TimeZero", options.timeZero);
+  alg.setProperty("LoadedTimeZero", options.loadedTimeZero);
+  alg.setProperty("Xmin", options.timeLimits.first);
+  double Xmax = options.timeLimits.second;
+  if (Xmax != Mantid::EMPTY_DBL()) {
+    alg.setProperty("Xmax", Xmax);
+  }
+
+  bool applyDTC = getProperty("ApplyDeadTimeCorrection");
+  if (applyDTC) {
+    TableWorkspace_sptr DTC = getProperty("DeadTimeTable");
+    alg.setProperty("ApplyDeadTimeCorrection", true);
+    alg.setProperty("DeadTimeTable", DTC);
+  }
+}
+
+// Set OutputType property of MuonProcess
+void ApplyMuonDetectorGrouping::setMuonProcessAlgorithmOutputTypeProperty(
+    IAlgorithm &alg, const Muon::AnalysisOptions &options) const {
+
+  std::string outputType;
+  switch (options.plotType) {
+  case Muon::PlotType::Counts:
+  case Muon::PlotType::Logarithm:
+    outputType = "GroupCounts";
+    break;
+  case Muon::PlotType::Asymmetry:
+    outputType = "GroupAsymmetry";
+    break;
+  default:
+    throw std::invalid_argument(
+        "Cannot create analysis workspace: Unsupported plot type");
+  }
+  alg.setProperty("OutputType", outputType);
+}
+
+// Set grouping properies of MuonProcess
+void ApplyMuonDetectorGrouping::setMuonProcessAlgorithmGroupingProperties(
+    IAlgorithm &alg, const Muon::AnalysisOptions &options) const {
+
+  alg.setProperty("DetectorGroupingTable", options.grouping.toTable());
+  alg.setProperty("GroupIndex", 0);
+}
+
+/**
+ * Set algorithm properties according to the given options. For use with
+ * MuonProcess.
+ */
+void ApplyMuonDetectorGrouping::setMuonProcessAlgorithmProperties(
+    IAlgorithm &alg, const Muon::AnalysisOptions &options) const {
+
+  alg.setProperty("Mode", "Combined");
+  alg.setProperty("CropWorkspace", false);
+  if (!options.rebinArgs.empty()) {
+    alg.setProperty("RebinParams", options.rebinArgs);
+  }
+  setMuonProcessAlgorithmGroupingProperties(alg, options);
+  setMuonProcessAlgorithmOutputTypeProperty(alg, options);
+  setMuonProcessAlgorithmTimeProperties(alg, options);
+}
+
+/**
```

Comment:
I would be tempted to move this nearer `init`, maybe underneath, so it's  init then basically validate inputs are closer together.